### PR TITLE
Dev merge modules part2

### PR DESCRIPTION
**Pull request by Flowingsun007**
添加module 以及 test case，包括：
- nn.Sigmoid
- nn.Softmax
- nn.LogSoftmax
- Tensor.masked_fill
- Tensor.unsqueeze
- Tensor.eq
- flow.arange

### CODE REVIEW

File: oneflow/python/nn/modules/activation.py [hunk 1]
```diff
@@ -164,3 +183,148 @@ def forward(self, x):
...
+        # [[-2.251349   -3.8766491  -0.13464898]
+        # [-0.48770458 -3.3176045  -1.0506046 ]]
+    """
+
+    def __init__(
+        self, dim: Optional[int] = 1,
+    ):
+        super().__init__()
+
+        self.dim = dim
```
Comment (Ldpe2G) [inline oneflow/python/nn/modules/activation.py:307]: 下面这三个 op 应该用 module

File: oneflow/python/nn/modules/eq.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        super().__init__()
+        self.eq_op = (
+            flow.builtin_op("broadcast_equal").Input("x").Input("y").Output("z").Build()
+        )
+
+    def forward(self, input, other):
+        for i in range(len(input.size())):
+            assert (
+                input.shape[i] >= other.shape[i]
+            ), "The second argument can be a number or a tensor whose shape is broadcastable with the first argument."
```
Comment (Ldpe2G) [inline oneflow/python/nn/modules/eq.py:33]: 这个注释是不是有问题，目前我们只支持 tensor 输入？

File: oneflow/python/nn/modules/eq.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        return self.eq_op(input, other)[0]
+
+
+@oneflow_export("eq")
+@register_tensor_op("eq")
+def eq_op(input, other):
+    r"""
+    Computes element-wise equality.
+    The second argument can be a number or a tensor whose shape is broadcastable with the first argument.
```
Comment (Ldpe2G) [inline oneflow/python/nn/modules/eq.py:42]: 同上

File: oneflow/python/nn/modules/eq.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        )
+
+    def forward(self, input, other):
+        for i in range(len(input.size())):
+            assert (
+                input.shape[i] >= other.shape[i]
+            ), "The second argument can be a number or a tensor whose shape is broadcastable with the first argument."
+        return self.eq_op(input, other)[0]
+
+
```
Comment (Ldpe2G) [inline oneflow/python/nn/modules/eq.py:36]: 可以同时多注册 equal 这个名字
https://pytorch.org/docs/stable/generated/torch.equal.html

File: oneflow/python/nn/modules/masked_fill.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+from oneflow.python.nn.module import Module
+from oneflow.python.oneflow_export import oneflow_export
+from oneflow.python.framework.tensor import register_tensor_op
+
+
+class MaskedFill(Module):
+    def __init__(self, value) -> None:
+        super().__init__()
+        self.value = value
+        self._where_op = (
```
Comment (Ldpe2G) [inline oneflow/python/nn/modules/masked_fill.py:26]: where 也应该单独做成一个 module

File: oneflow/python/nn/modules/unsqueeze.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+from oneflow.python.nn.module import Module
+from oneflow.python.oneflow_export import oneflow_export
+from oneflow.python.framework.tensor import register_tensor_op
+
+
+class Unsqueeze(Module):
+    def __init__(self, dim: int = 0,) -> None:
+        super().__init__()
+        self.dim = dim
+        self._op = flow.builtin_op("expand_dims").Input("in").Output("out").Build()
```
Comment (Ldpe2G) [inline oneflow/python/nn/modules/unsqueeze.py:26]: expand_dim 在https://github.com/Oneflow-Inc/oneflow/pull/4775 这个 pr 已经添加为 module 了

File: oneflow/python/nn/modules/eq.py (hunk 2 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/eq.py:33]: 这个注释是broadcast_equal这个op的docs里的注释

File: oneflow/python/nn/modules/activation.py (hunk 1 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/activation.py:380]: 好的

File: oneflow/python/nn/modules/eq.py (hunk 4 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/eq.py:36]: 已加equal注册

File: oneflow/python/nn/modules/unsqueeze.py (hunk 6 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/unsqueeze.py:26]: 这个pr里只有expand module吧(基于expand op实现的，和expand_dim op的用法有一点区别）

File: oneflow/python/nn/modules/masked_fill.py (hunk 5 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/masked_fill.py:26]: 先mark一下。这个where依赖[broadcast_to_compatible_with](https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/python/ops/math_ops.py#L1566)的op，是opconf拼出来的（非userop），暂时不好搭module

File: oneflow/python/nn/modules/activation.py [hunk 7]
```diff
@@ -164,3 +183,218 @@ def forward(self, x):
...
+            x = x.transpose(perm=permute)
+
+        res = self._op(x)[0]
+        if need_transpose:
+            res = res.transpose(perm=permute)
+        return res
+
+
+@oneflow_export("softmax")
+def softmax_op(tensor, /, dim=None):
```
Comment (Ldpe2G) [inline oneflow/python/nn/modules/activation.py:286]: / 应该是先不要加，不是说 python 3.8 之后才支持

File: oneflow/python/nn/modules/unsqueeze.py (hunk 6 above)
Comment (Ldpe2G) [reply oneflow/python/nn/modules/unsqueeze.py:26]: 看错了

File: oneflow/python/nn/modules/eq.py (hunk 2 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/eq.py:33]: 嗯，目前只支持tensor输入，我改下

File: oneflow/python/nn/modules/activation.py (hunk 7 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/activation.py:286]: ok,那我去掉先

File: oneflow/python/nn/modules/unsqueeze.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+@oneflow_export("unsqueeze")
+@register_tensor_op("unsqueeze")
+def unsqueeze_op(tensor, dim):
+    r"""Returns a new tensor with a dimension of size one inserted at the
+    specified position.
+
+    The returned tensor shares the same underlying data with this tensor.
+
+    A :attr:`dim` value within the range ``[0, input.dim() + 1)``
```
Comment (BBuf) [inline oneflow/python/nn/modules/unsqueeze.py:43]: 这里和Pytorch注释一样但是上面的forward里面把dim=-xxx的情况屏蔽了，这里需要改一下。

File: oneflow/python/test/modules/test_activation.py [hunk 9]
```diff
@@ -119,5 +119,225 @@ def test_gelu_v3(test_case):
...
+
+
+@unittest.skipIf(
+    not flow.unittest.env.eager_execution_enabled(),
+    ".numpy() doesn't work in lazy mode",
+)
+class TestSoftmaxModule(flow.unittest.TestCase):
+    def test_softmax(test_case):
+        m = flow.nn.Softmax(dim=1)
+        x = flow.Tensor(
```
Comment (BBuf) [inline oneflow/python/test/modules/test_activation.py:155]: 这里可以直接random生成数据，感觉没必要把值打出来，不太好看。

File: oneflow/python/test/modules/test_activation.py [hunk 10]
```diff
@@ -119,5 +119,225 @@ def test_gelu_v3(test_case):
...
+        test_case.assertTrue(np.allclose(y.numpy(), output, rtol=1e-05))
+
+    def test_softmax_dim_2(test_case):
+        x = flow.Tensor(
+            np.array(
+                [
+                    [
+                        [
+                            [-0.46716809, 0.40112534, 0.61984003],
+                            [-1.31244969, -0.42528763, 1.47953856],
```
Comment (BBuf) [inline oneflow/python/test/modules/test_activation.py:186]: 同上

File: oneflow/python/test/modules/test_activation.py [hunk 11]
```diff
@@ -119,5 +119,225 @@ def test_gelu_v3(test_case):
...
+            )
+        )
+
+        y = m(x)
+        output = np.array(
+            [
+                [
+                    [
+                        [-0.22100820, -0.30664772, -1.50251734, -2.31782818],
+                        [-2.05509090, -1.53121281, -1.31776488, -0.32986164],
```
Comment (BBuf) [inline oneflow/python/test/modules/test_activation.py:315]: 同上，因为这些都是numpy可以直接做到的，感觉就不需要把值打出来。

File: oneflow/python/test/modules/test_unsqueeze.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+class TestUnsqueeze(flow.unittest.TestCase):
+    def test_unsqueeze(test_case):
+        x = flow.Tensor(np.random.rand(2, 6, 9, 3))
+        y = flow.unsqueeze(x, dim=1)
+        test_case.assertTrue(np.allclose(flow.Size([2, 1, 6, 9, 3]), y.shape))
+
+    def test_unsqueeze_tensor_function(test_case):
+        x = flow.Tensor(np.random.rand(2, 3, 4))
+        y = x.unsqueeze(dim=2)
+        test_case.assertTrue(np.allclose(flow.Size([2, 3, 1, 4]), y.shape))
```
Comment (BBuf) [inline oneflow/python/test/modules/test_unsqueeze.py:35]: 需要增加dim为负数的测试样例。

File: oneflow/python/nn/modules/unsqueeze.py (hunk 8 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/unsqueeze.py]: 已修改

File: oneflow/python/nn/modules/activation.py [hunk 13]
```diff
@@ -164,3 +183,218 @@ def forward(self, x):
...
+        super().__init__()
+        self._op = flow.builtin_op("sigmoid").Input("in").Output("out").Build()
+
+    def forward(self, x):
+        return self._op(x)[0]
+
+
+@oneflow_export("sigmoid")
+@register_tensor_op("sigmoid")
+def sigmoid_op(tensor):
```
Comment (BBuf) [inline oneflow/python/nn/modules/activation.py:232]: 根据文档来看，把这个变量改成x吧，这样对用户好看一点。

File: oneflow/python/nn/modules/unsqueeze.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        ), "dim should within the range [-input.ndimension() - 1, input.ndimension() + 1)"
+        
+        if self.dim < 0:
+            self.dim = 1+input.ndimension() + self.dim
+        return self._op(input, axis=self.dim)[0]
+
+
+@oneflow_export("unsqueeze")
+@register_tensor_op("unsqueeze")
+def unsqueeze_op(tensor, dim):
```
Comment (BBuf) [inline oneflow/python/nn/modules/unsqueeze.py:40]: 根据文档来看，把这个变量改成input吧，这样对用户好看一点。

File: oneflow/python/test/modules/test_activation.py [hunk 15]
```diff
@@ -119,5 +119,169 @@ def test_gelu_v3(test_case):
...
 
 
+@unittest.skipIf(
+    not flow.unittest.env.eager_execution_enabled(),
+    ".numpy() doesn't work in lazy mode",
+)
+class TestSigmoidModule(flow.unittest.TestCase):
+    def test_sigmoid(test_case):
+        m = flow.nn.Sigmoid()
+        x = flow.Tensor(
```
Comment (BBuf) [inline oneflow/python/test/modules/test_activation.py:129]: 这里也改一下吧，sigmoid的numpy实现：

```suggestion
def sigmoid(x):
  return 1 / (1 + math.exp(-x))
```

File: oneflow/python/test/modules/test_activation.py [hunk 16]
```diff
@@ -119,5 +119,169 @@ def test_gelu_v3(test_case):
...
+
+    def test_logsoftmax_v2(test_case):
+        m = flow.nn.LogSoftmax(dim=2)
+        x = flow.Tensor(
+            np.array(
+                [
+                    [
+                        [
+                            [2.55630851, 1.57471120, 0.25240266, -0.57634264],
+                            [0.72222596, 0.35014620, 0.43715513, 1.41162395],
```
Comment (BBuf) [inline oneflow/python/test/modules/test_activation.py:228]: logsoftmax的测试也要改改。

logsoftmax的numpy实现大概长这样子：

```python
def log_softmax(x, dim):
    e_x = np.exp(x - np.max(x, axis=dim, keepdims=True))
    return np.log(e_x / e_x.sum(axis=dim, keepdims=True))
```

File: oneflow/python/test/modules/test_eq.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        arr1 = np.array([2, 3, 4, 5,])
+        arr2 = np.array([2, 3, 4, 1])
+        input = flow.Tensor(arr1, dtype=flow.float32)
+        other = flow.Tensor(arr2, dtype=flow.float32)
+
+        of_out = flow.eq(input, other)
+        of_out2 = flow.equal(input, other)
+        np_out = np.equal(arr1, arr2)
+        test_case.assertTrue(np.allclose(of_out.numpy(), np_out))
+        test_case.assertTrue(np.allclose(of_out2.numpy(), np_out))
```
Comment (BBuf) [inline oneflow/python/test/modules/test_eq.py:37]: allclose改成array_equal

File: oneflow/python/test/modules/test_masked_fill.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+                    [0.34675095, -0.7050969, -0.7647731, -0.73233418],
+                    [-1.90089858, 0.01262963, 0.74693893, 0.57132389],
+                ],
+            ]
+        )
+
+        output = np.array(
+            [
+                [
+                    [-0.1316931, 8.7654321, 8.7654321, 8.7654321],
```
Comment (BBuf) [inline oneflow/python/test/modules/test_masked_fill.py:46]: 这里应该也可以改，numpy实现了这个OP，参考这里：https://numpy.org/doc/stable/reference/generated/numpy.ma.filled.html

File: oneflow/python/test/modules/test_unsqueeze.py [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+        y = flow.unsqueeze(x, dim = 1)
+        output = np.expand_dims(np_arr, axis = 1)
+        test_case.assertTrue(np.allclose(output, y.numpy(), rtol=1e-05))
+
+    def test_unsqueeze_tensor_function(test_case):
+        np_arr = np.random.rand(2, 3, 4)
+        x = flow.Tensor(np_arr)
+        y = x.unsqueeze(dim = 2)
+        output = np.expand_dims(np_arr, axis = 2)
+        test_case.assertTrue(np.allclose(output, y.numpy(), rtol=1e-05))
```
Comment (BBuf) [inline oneflow/python/test/modules/test_unsqueeze.py:39]: allclose可以改成array_equal，这里不存在精度损失

Comment (Flowingsun007) [reply oneflow/python/test/modules/test_unsqueeze.py:39]: 应该还是有精度损失的（试了用array_equal会报错

Comment (BBuf) [reply oneflow/python/test/modules/test_unsqueeze.py:39]: 那还是改回来吧，我猜测是numpy默认用了float64的原因

File: oneflow/python/test/modules/test_masked_fill.py (hunk 18 above)
Comment (Flowingsun007) [reply oneflow/python/test/modules/test_masked_fill.py:46]: 这个用np.ma好像不太好实现，感觉也没啥必要-_-

File: oneflow/python/test/modules/test_eq.py (hunk 17 above)
Comment (Flowingsun007) [reply oneflow/python/test/modules/test_eq.py]: 已修改

File: oneflow/python/test/modules/test_activation.py (hunk 16 above)
Comment (Flowingsun007) [reply oneflow/python/test/modules/test_activation.py]: 已修改

File: oneflow/python/test/modules/test_masked_fill.py (hunk 18 above)
Comment (BBuf) [reply oneflow/python/test/modules/test_masked_fill.py:46]: 也可以，那保留这种做法吧

**Review by BBuf — APPROVED** [review]

File: oneflow/python/nn/modules/eq.py [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+class Eq(Module):
+    def __init__(self) -> None:
+        super().__init__()
+        self.eq_op = (
+            flow.builtin_op("broadcast_equal").Input("x").Input("y").Output("z").Build()
+        )
+
+    def forward(self, input, other):
+        if isinstance(other, flow.Tensor):
```
Comment (Ldpe2G) [inline oneflow/python/nn/modules/eq.py:30]: 可以加一个 assert len(input.shape) == len(other.shape)

**Review by Ldpe2G — APPROVED** [review]
